### PR TITLE
Refactor function code and allow for using starlark based filters

### PR DESCRIPTION
**Pull request by pwittrock**

### CODE REVIEW

File: cmd/config/internal/commands/run-fns.go [hunk 1]
```diff
@@ -68,31 +81,63 @@ func (r *RunFnRunner) runE(c *cobra.Command, args []string) error {
...
+			if err != nil {
+				return nil, err
+			}
+		}
 	}
-	if r.Network {
+
+	if r.EnableStar && r.StarPath != "" {
+		// if image isn't specified, then Functions is empty
+		if r.StarPath == "" {
```
Comment (phanimarupaka) [inline cmd/config/internal/commands/run-fns.go]: Does this ever happen as the upper if checks r.StarPath != ""

File: cmd/config/internal/commands/run_test.go [hunk 2]
```diff
@@ -154,6 +154,34 @@ kind: Foo
...
+    config.kubernetes.io/function: |
+      starlark: {path: a/b/c, name: foo}
+data: {g: h}
+kind: Foo
+apiVersion: v1
+`,
+		},
+		{
+			name: "star-not-enabled",
+			args: []string{"run", "dir",
```
Comment (phanimarupaka) [inline cmd/config/internal/commands/run_test.go]: Maybe add another case like this 

[]string{"run", "dir",
                                 "--image", "some_image",
				"--star-path", "a/b/c",
				"--star-name", "foo",
				"--", "Foo", "g=h"},

File: cmd/config/internal/commands/run-fns.go (hunk 1 above)
Comment (pwittrock) [reply cmd/config/internal/commands/run-fns.go]: Nope

File: cmd/config/internal/commands/run_test.go (hunk 2 above)
Comment (pwittrock) [reply cmd/config/internal/commands/run_test.go]: done

File: cmd/config/internal/commands/run_test.go [hunk 3]
```diff
@@ -152,6 +152,53 @@ metadata:
...
+		{
+			name: "image-star-not-enabled",
+			args: []string{"run", "dir",
+				"--image", "some_image",
+				"--star-path", "a/b/c",
+				"--star-name", "foo",
+				"--", "Foo", "g=h"},
+			path: "dir",
+			expected: `
+metadata:
```
Comment (phanimarupaka) [inline cmd/config/internal/commands/run_test.go]: I think this should throw an error something like "star-path specified but enable-star is not set". I don't think we should silently accept the arguments and not honor them.

File: kyaml/runfn/runfn.go [hunk 4]
```diff
@@ -57,6 +58,12 @@ type RunFns struct {
 	// and only use explicit sources
 	NoFunctionsFromInput *bool
 
+	// EnableStarlark will enable functions run as starlark scripts
+	EnableStarlark bool
+
+	// DisableContainers will disable functions run as containers
+	DisableContainers bool
```
Comment (monopole) [inline kyaml/runfn/runfn.go]: This is a boolean mess.

Comment (monopole) [reply kyaml/runfn/runfn.go]: can we do an enumeration of legal states instead?

Comment (monopole) [reply kyaml/runfn/runfn.go]: or are all four states legal?  do we add more booleans for _cue_ or whatever's next?  
I've read the PR description, but it's not helping me.

Comment (pwittrock) [reply kyaml/runfn/runfn.go]: - yes all 4 states are legal -- you can enable starlark only, containers only, starlark and containers, neither starlark or containers.  The last state is legal but not useful.
- booleans map pretty well to command line flags, enumerations less so
- most folks should be using the defaults

This api doesn't have a v1.0.0 release, so it is pretty easy to change.  I'd suggest we keep this as is and refactor when it is more obvious what a better representation would be.

**Pull request merged**